### PR TITLE
Cameo4 support

### PR DESCRIPTION
**Pull request by mr-sven**
couldn't resist to clean up the code

### CODE REVIEW

**Comment by mr-sven** [comment]
next step should be to eliminate all magical milimeter/inch to SilhuetteUnit conversions

File: silhouette/Graphtec.py [hunk 1]
```diff
@@ -573,216 +699,196 @@ def setup(s, media=132, speed=None, pressure=None, toolholder=None, pen=None, cu
...
       # This value is the circle diameter which is exectuted on direction changes on corners to adjust the blade.
       # Seems to be limited to 46 or 47. Values above does keep the last setting on the device.
-      if s.product_id() == PRODUCT_ID_SILHOUETTE_CAMEO3:
+      if self.product_id() == PRODUCT_ID_SILHOUETTE_CAMEO3:
         if not pen:
-          circle = 0.5 + bladediameter * 20
-          s.write(b"FC0,1,%d\x03FC%d,1,%d\x03" % (toolholder, circle, toolholder))
+          self.send_command([
+            tool.cutter_offset(0, 0.05),
+            tool.cutter_offset(bladediameter + 0.05, 0.05)])
```
Comment (FlyingSamson) [inline silhouette/Graphtec.py]: I think the new dx is differently rounded then the old one:
For instance for bladediameter=0.85 we get:
Old: (0.5+0.85*20)SU=17.5 SU -> as int 17SU
New: (0.85+0.05)*20 SU = 18 SU -> as int still 18SU
If we want to maintain the rounding to the closest number in SUs the correct formula would be `bladediameter + 0.025` correct?
Please correct me, if I don't see why the new way of rounding is superior to the old one.

File: silhouette/Graphtec.py [hunk 2]
```diff
@@ -1036,29 +1142,32 @@ def plot(s, mediawidth=210.0, mediaheight=297.0, margintop=None,
...
-      p += b"\x03D%d,%d" % (int(0.5+bbox['ury']), int(0.5+bbox['urx']))
-      p += b"\x03D%d,%d" % (int(0.5+bbox['lly']), int(0.5+bbox['urx']))
-      p += b"\x03D%d,%d" % (int(0.5+bbox['lly']), int(0.5+bbox['llx']))
-      p += b"\x03D%d,%d" % (int(0.5+bbox['ury']), int(0.5+bbox['llx']))
-    p += b"\x03"   # Properly terminate string of plot commands.
+      cmd_list.append("M%d,%d" % (int(0.5+bbox['ury']), int(0.5+bbox['llx'])))
+      cmd_list.append("D%d,%d" % (int(0.5+bbox['ury']), int(0.5+bbox['urx'])))
+      cmd_list.append("D%d,%d" % (int(0.5+bbox['lly']), int(0.5+bbox['urx'])))
+      cmd_list.append("D%d,%d" % (int(0.5+bbox['lly']), int(0.5+bbox['llx'])))
+      cmd_list.append("D%d,%d" % (int(0.5+bbox['ury']), int(0.5+bbox['llx'])))
```
Comment (FlyingSamson) [inline silhouette/Graphtec.py]: The current implementation overrides `p` here, so as to discard the command list obtained with `self.plot_cmds`, as we apparently want to draw only the bbox. So I presume the first line should be `cmd_list =...` or an additional `cmd_list = []` would be needed.

File: silhouette/Graphtec.py (hunk 1 above)
Comment (mr-sven) [reply silhouette/Graphtec.py]: It took my a time to identify the idea behind the 0.5, but now its clear. But it should be `0.0025`

File: silhouette/Graphtec.py (hunk 2 above)
Comment (mr-sven) [reply silhouette/Graphtec.py]: ok, I see it.

File: silhouette/Graphtec.py (hunk 1 above)
Comment (FlyingSamson) [reply silhouette/Graphtec.py]: Hm, no I think is should be 0.025, as then
```[new code] = (int)((bladediameter + 0.025) * 20) = (int)(0.025 * 20 + bladediameter * 20) = (int)(0.5 + bladediameter * 20) = [old code]```

File: silhouette/Graphtec.py [hunk 3]
```diff
@@ -627,45 +620,31 @@ def initialize(self):
...
     if self.dev is None: return None
 
-    self.send_command("FG")   # Get Version: 17 Chars
-    try:
-      resp = self.read(timeout=10000) # Large timeout because the plotter moves.
-    except usb.core.USBError as e:
-      print("usb.core.USBError:", e, file=self.log)
-      resp = "None  "
-    return resp[0:-2]   # chop of 0x03
+    return self.send_receive_command("FG", rx_timeout = 10000)
```
Comment (FlyingSamson) [inline silhouette/Graphtec.py:694]: Here you are chopping one character (0x03) instead of two in the previous implementation. However, I don't see why we would want to remove two tailing characters here. So am I right in assuming, that you didn't saw a reason for removing the last two characters either and therefore changed the behavior accordingly, or am I overseeing something and we should indeed remove two characters?

File: silhouette/Graphtec.py [hunk 4]
```diff
@@ -138,11 +138,13 @@
 # Negative Acnoledge - Returns device tool setup
 CMD_NAK = '\x15'
 
+SILHOUETTE_CAMEO4_TOOL_EMPTY = 0
 SILHOUETTE_CAMEO4_TOOL_BLADE = 1
 SILHOUETTE_CAMEO4_TOOL_AUTOBLADE = 2
 SILHOUETTE_CAMEO4_TOOL_DEEPCUT = 3
 SILHOUETTE_CAMEO4_TOOL_2MM_CRAFTBLADE = 4
```
Comment (FlyingSamson) [inline silhouette/Graphtec.py]: You can further add
```
SILHOUETTE_CAMEO4_TOOL_ROTARYBLADE = 5
```
here.

Also it might make sense to rename `DEEPCUT`to `DEEPCUTBLADE` to be consistent with `KRAFTBALDE` and `AUTOBLADE`

Comment (mr-sven) [reply silhouette/Graphtec.py]: Check

Comment (FlyingSamson) [inline silhouette/Graphtec.py]: Furthermore I hava a 3mm Kraftblade, and it is also reported back by the Cameo4 as `4`, so I would just write
```
SILHOUETTE_CAMEO4_TOOL_KRAFTBLADE = 4
```

Comment (mr-sven) [reply silhouette/Graphtec.py]: Ok, so its only kraftblade

Comment (FlyingSamson) [reply silhouette/Graphtec.py]: exactly

File: silhouette/Graphtec.py (hunk 3 above)
Comment (mr-sven) [reply silhouette/Graphtec.py:694]: Did you see any response with two times of a 0x03? I had no idea why there was a chop of two chars

Comment (mr-sven) [reply silhouette/Graphtec.py:694]: I could replace it with `.rstrip(CMD_ETX)`

Comment (FlyingSamson) [reply silhouette/Graphtec.py:694]: No, I did not. That was why I was confused by the old code. I presume it was just a typo or a blackout by the writer, where he or she thought that 0x03 must be two chars '0' and '3', or something. So I guess it makes more sense to chop the one character as also indicated by the comment.

I will mark this as resolved, just wanted to make sure that the change of behavior was intended and not a mistake.

File: silhouette/Graphtec.py (hunk 1 above)
Comment (mr-sven) [reply silhouette/Graphtec.py]: check

File: silhouette/Graphtec.py [hunk 5]
```diff
@@ -842,6 +842,7 @@ def setup(self, media=132, speed=None, pressure=None, toolholder=None, pen=None,
       if pen:
         self.send_command(tool.sharpen_corners(0, 0))
       else:
+        # TODO: shouldn't be this also SU? why * 10 ?
```
Comment (FlyingSamson) [inline silhouette/Graphtec.py]: I think that times 10 was right (also I don't have a clue as to why they don't use SUs here), but I will have another look at it with wireshark and will report back on my findings.

File: silhouette/Graphtec.py [hunk 6]
```diff
@@ -1040,6 +1040,11 @@ def plot_cmds(self, plist, bbox, x_off_mm, y_off_mm):
     # As a result we get varying offsets which can be really annoying if doing precision
     # work.
 
+    # Change by Sven Fabricius:
+    # Update the code to use millimetres in all places to prevent mixing with device units.
+    # The conversation to SU (SilhuetteUnits) will be done in command create function.
```
Comment (FlyingSamson) [inline silhouette/Graphtec.py]: I presume you mean conversion here

File: silhouette/Graphtec.py [hunk 7]
```diff
@@ -141,8 +141,9 @@
 SILHOUETTE_CAMEO4_TOOL_EMPTY = 0
 SILHOUETTE_CAMEO4_TOOL_BLADE = 1
```
Comment (FlyingSamson) [inline silhouette/Graphtec.py]: I think 1 is Ratchet blade (black adapter), so it might make sense to rename it correspondingly to `RATCHETBLADE`. Or is there another blade from Cameo3 that is reported as 1?

File: silhouette/Graphtec.py (hunk 5 above)
Comment (FlyingSamson) [reply silhouette/Graphtec.py]: So I gave it another look and indeed the numbers are in 1/10th of millimeter, i.e. 0.1mm translates to 1, 0.5mm translates to 5, and so on.
To my surprise the FF$1,$2,$3 command behaves similar to the TJ command:
* $1 is updated immediately, as in I change the start ext. in SS (e.g. 0.1 to 0.5) and it will be transmitted accordingly in the next send command for both FF_,_,_ commands
* BUT $2 differs between the two FF_,_,_ commands: The first command repeats the old value (e.g. 0.1) and the second holds the new value (e.g. 0.5)

And as with the TJ command in the first send after reconnection of the Cameo it sends 0 as old value (probably because the old value is stored per connection or something). With this in mind, I think you might have been right about TJ being acceleration after all. I will investigate this further.

To sum things up times 10 here is correct, so I think you should alter that comment to state that 10 is correct to prevent future code reviewers from wondering about that and checking it again and again.

File: silhouette/Graphtec.py [hunk 8]
```diff
@@ -458,331 +619,355 @@ def wait_for_ready(s, timeout=30, verbose=True):
...
         if pressure > 33: pressure = 33
-        s.write(b"FX%d,%d\x03" % (pressure, toolholder));
-        print("pressure: %d" % pressure, file=s.log)
+        self.send_command(tool.pressure(pressure))
+        print("pressure: %d" % pressure, file=self.log)
+
         # some sort of pressure calibration I presume (only 0 on first connection to cameo else 3)
-        # hopefully also allowed later on
-        s.write(b"TJ0\x03")
+        # hopefully also allowed later on, maybe accelleration
```
Comment (FlyingSamson) [inline silhouette/Graphtec.py]: You are right off course TJ transmits the acceleration for Cameo4. But as for the FF_,_,_ command it first sends the previously used value (0 after first connection) and then the new value used for the upcoming plot. So I guess as with the FF we should keep the double sending of the command and also as with FF always send 0 with the first command as if it was reconnected.
